extractor/main_test.go

```diff
@@ -0,0 +1,412 @@
+		require.NoError(t, err)
+
+		// Read file
+		file, err := os.Open(filepath.Join(outputDir, packageMetadataFile))
+		require.NoError(t, err)
+
+		reader := bufio.NewReader(file)
+		var unmarshalledMetadata Metadata
+
+		raw, isPrefix, err := reader.ReadLine()
+		require.NoError(t, err)
+
+		if isPrefix {
+			t.Fatalf("Metadata longer then buffer max size at %s\n", outputDir)
+		}
+
+		err = json.Unmarshal(raw, &unmarshalledMetadata)
+		require.NoError(t, err)
+
+		t.Cleanup(func() {
+			err := os.RemoveAll(tempDir)
+			if err != nil {
+				fmt.Printf("could not clean up temp dir, %v", err)
+				return
+			}
+		})
+
+		require.Equal(t, md, unmarshalledMetadata)
+	}
+}
+
+func TestWritePackageFiles(t *testing.T) {
+	t.Parallel()
+
+	_, mockMsgsAddPackage := generateMockMsgs(t)
+
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, msg := range mockMsgsAddPackage {
+		// Get output dir
+		outputDir := filepath.Join(tempDir, strings.TrimLeft(msg.Package.Path, "gno.land/"))
+
+		// Write dir before writing metadata
+		if dirWriteErr := os.MkdirAll(outputDir, os.ModePerm); dirWriteErr != nil {
+			t.Fatal(dirWriteErr)
+		}
+
+		// Write the metadata
+		err := writePackageFiles(msg, outputDir)
+		require.NoError(t, err)
+
+		// Read & compare file
+		for _, f := range msg.Package.Files {
+			contents, err := os.ReadFile(filepath.Join(outputDir, f.Name))
+			require.NoError(t, err)
+			require.Equal(t, f.Body, string(contents))
+		}
+	}
+	t.Cleanup(func() {
+		err := os.RemoveAll(tempDir)
+		if err != nil {
+			fmt.Printf("could not clean up temp dir, %v", err)
+			return
+		}
+	})
+}
+
+// Helpers
+func generateSourceFiles(t *testing.T, mockMsgs []std.Msg) []string {
+	t.Helper()
+
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		mockTx    = make([]std.Tx, numTx)
+		testFiles = make([]string, numSourceFiles)
+	)
+
+	// Generate transactions to wrap messages
+	for i := range mockTx { // num
+		mockTx[i] = std.Tx{
+			Msgs: mockMsgs[:msgPerTx],
+		}
+		mockMsgs = mockMsgs[msgPerTx:]
+	}
+
+	// Generate source file names
+	for i := 0; i < numSourceFiles; i++ {
+		testFiles[i] = "sourceFile" + strconv.Itoa(i) + sourceFileType
+	}
+
+	// Generate source files
+	for _, file := range testFiles {
+		filePath := filepath.Join(tempDir, file)
+		if err := os.MkdirAll(filepath.Dir(filePath), os.ModePerm); err != nil {
+			t.Fatal(err)
+		}
+		file, err := os.Create(filePath)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, tx := range mockTx[:txPerSourceFile] {
+
+			err := writeTxToFile(tx, file, t)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+		mockTx = mockTx[txPerSourceFile:]
+	}
+
+	for i := 0; i < numSourceFiles; i++ {
+		testFiles[i] = filepath.Join(tempDir, testFiles[i])
+	}
+
+	t.Cleanup(func() {
+		err := os.RemoveAll(tempDir)
+		if err != nil {
+			fmt.Printf("could not clean up temp dir, %v", err)
+			return
+		}
+	})
```

Comment:
What is the point of this cleanup, if you need the file paths outside the method?

Reply:
This cleanup should live outside

Reply:
It will get removed with the testing context, the point I was trying to make is you should centralize your resource management. The tests you wrote jumble this through the main execution body, but also through helpers (you manage resources in both), instead of for example creating this high-level testdir in the main test body, and just passing the path to the helper (for sub-files)

Reply:
This has actually been a problem when parallelizing tests. I have fixed it by moving the temporary dir management to the top level test function, instead of having it in the helpers. [3965302](https://github.com/gnolang/tx-exports/pull/5/commits/3965302f90c0f5773f9a0b05cadb3e66670b0028)